### PR TITLE
[papi] Extract JWT cookie when extracting creds

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

With this change, we no longer blindly extract all Cookies from the incoming request. Instead, we look for the JWT cookie by name, and attempt to validate it, before accepting the request.

This change has the benefit that for most Dashboard interactions, changes to the user's cookies (or even new cookies being added) won't actually invalidate their connection cache

## Related Issue(s)

Fixes #

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - mp-papi-exbfa6496e39</li>
	<li><b>🔗 URL</b> - <a href="https://mp-papi-exbfa6496e39.preview.gitpod-dev.com/workspaces" target="_blank">mp-papi-exbfa6496e39.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/public-api-server/pkg/auth/middleware.go [hunk 1]
```diff
@@ -53,40 +59,36 @@ func (a *Interceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) co
...
-	cookie := conn.RequestHeader().Get("Cookie")
-	if cookie != "" {
-		return NewCookieToken(cookie), nil
+	_, err = VerifySessionJWT(cookie.Value, i.verifier, i.sessionCfg.Issuer)
+	if err != nil {
+		return Token{}, connect.NewError(connect.CodeUnauthenticated, fmt.Errorf("JWT session could not be verified."))
 	}
 
-	return Token{}, connect.NewError(connect.CodeUnauthenticated, fmt.Errorf("No access token or cookie credentials available on request."))
+	return NewCookieToken(cookie.String()), nil
```
Comment (easyCZ) [inline components/public-api-server/pkg/auth/middleware.go:91]: This is the primary change, remainder are tests and/or cleanup

**Comment by geropl** [comment]
Starting review now...

File: components/public-api-server/pkg/auth/context.go [hunk 2]
```diff
@@ -23,7 +23,9 @@ const (
 )
 
 type Token struct {
-	Type  TokenType
+	Type TokenType
+	// When Type is AccessTokenType, the value is the raw token value
+	// When Type is CookieTokenType, the value is cookie_name=cooke_value
```
Comment (geropl) [inline components/public-api-server/pkg/auth/context.go:28]: Not really important, but: Di you considered making this Type, Value string, CookieName string (optional)? :thinking: Motivation being that clients don't have to parse Value over and over again. But not sure how it's used, so: leaving it up to you :wink:

**Comment by geropl** [comment]
Build:
```
[components/public-api-server:docker] error pulling image configuration: download failed after attempts=1: error parsing HTTP 499 response body: invalid character '<' looking for beginning of value: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><Error><Code>Unexpected</Code><Message>Client Disconnect</Message></Error>"
```
Me:
```
Re-run all jobs
```
:golfing_man:

**Pull request merged**